Implement CL_VERSION_* features

vmx opened the pull request:
`openccl3` specifies compile time features for the different OpenCL versions,
but they weren't fully implemented. This PR implements them. The functions
become the `CL_VERSION_*` number when they were introduced. So if you e.g.
want to use OpenCL 1.2 and 2.0 features, you would need to specify the
`CL_VERSION_1_2` and `CL_VERSION_2_0`. All features prior to OpenCL 1.2
are always enabled.

This commit enables you to compile a `opencl3` version with only OpenCL 1.2
features (without any from newer version), which would then run on platforms
that only support OpenCL 1.2 (like MacOS).

Trying out if it can actually built with OpenCL 1.2 features only can be done with:

    cargo build --no-default-features --features CL_VERSION_1_2 --all-targets

It was manually tested that compiling with each `CL_VERSION_*` feature
individually, while disabling the other ones, works.

Fixes #30.

vmx commented:
For easier review, hide the whitespace changes: https://github.com/kenba/opencl3/pull/31/files?diff=split&w=1.

vmx commented:
I'm happy to adjust the PR, I just want to mention that this PR worked for my use case, I just verified on CI :) You can include `opencl3` in your library and it won't include symbols that are not part of OpenCL 1.2 (I didn't verify in detail, but it at least doesn't contain the symbols current master would include).

kenba commented on src/command_queue.rs:
```diff
@@ -1240,19 +1255,20 @@ impl CommandQueue {
         Ok(get_command_queue_info(self.queue, CommandQueueInfo::CL_QUEUE_PROPERTIES)?.to_ulong())
     }
 
+    #[cfg(feature = "CL_VERSION_2_0")]
```
I think that this is incorrect.  
The call to `get_command_queue_info` should be in every OpenCL ICD, it is not an "CL_VERSION_2_0" feature.  
Therefore, since `CL_QUEUE_PROPERTIES` is just an attribute of an OpenCL 2 device it should be accessible from *any* OpenCL ICD and **not** an "CL_VERSION_2_0" feature.

Perhaps it should just be commented as CL_VERSION_2_1 was on line 1247 below.

kenba commented on src/command_queue.rs:
```diff
@@ -1240,19 +1255,20 @@ impl CommandQueue {
     }
 
-    // CL_VERSION_2_1
+    #[cfg(feature = "CL_VERSION_2_1")]
```
As above.

kenba commented on src/command_queue.rs:
```diff
@@ -1240,19 +1255,20 @@ impl CommandQueue {
     }
 
-    // CL_VERSION_3_0
+    #[cfg(feature = "CL_VERSION_3_0")]
```
Again, as above.

kenba commented on src/command_queue.rs:
```diff
@@ -1268,6 +1284,7 @@ mod tests {
     use crate::device::Device;
     use crate::platform::get_platforms;
     use cl3::device::CL_DEVICE_TYPE_GPU;
+    #[cfg(feature = "CL_VERSION_2_1")]
```
I'd prefer `#[allow(unused_imports)]` than requiring a feature for a type.

kenba commented on src/command_queue.rs:
```diff
@@ -1307,19 +1324,19 @@ mod tests {
         // assert_eq!(2, value);
 
-        // CL_VERSION_2_0 value
+        #[cfg(feature = "CL_VERSION_2_0")]
```
As above, the call to `queue.size()` should not be an CL_VERSION_2_0 feature, therefore this change is unnecessary.

kenba commented on src/command_queue.rs:
```diff
@@ -1307,19 +1324,19 @@ mod tests {
         }
 
-        // CL_VERSION_2_1 value
+        #[cfg(feature = "CL_VERSION_2_1")]
```
As previous comment.

kenba commented on src/command_queue.rs:
```diff
@@ -1307,19 +1324,19 @@ mod tests {
         }
 
-        // CL_VERSION_3_0 value
+        #[cfg(feature = "CL_VERSION_3_0")]
```
As previous comments.

kenba commented on src/context.rs:
```diff
@@ -197,6 +200,7 @@ impl Context {
 
     /// Get the common Shared Virtual Memory (SVM) capabilities of the
     /// devices in the Context.
+    #[cfg(feature = "CL_VERSION_2_0")]
```
This is a property of  device, it is **not** a CL_VERSION_2_0 feature.

kenba commented on src/context.rs:
```diff
@@ -348,6 +356,7 @@ mod tests {
         let device = Device::new(devices[0]);
         let context = Context::from_device(&device).unwrap();
 
+        #[cfg(feature = "CL_VERSION_2_0")]
```
Not a CL_VERSION_2_0 feature.

kenba commented on src/device.rs:
```diff
@@ -16,9 +16,14 @@ pub use cl3::device::*;
 pub use cl3::ffi::cl_ext::{cl_amd_device_topology, CL_LUID_SIZE_KHR, CL_UUID_SIZE_KHR};
 
 use super::Result;
+#[cfg(feature = "CL_VERSION_1_2")]
```
Please use `#[allow(unused_imports)]` instead of specifying specific types for versions.

kenba commented on src/device.rs:
```diff
@@ -345,6 +355,7 @@ impl Device {
         Ok(get_device_info(self.id(), DeviceInfo::CL_DEVICE_PLATFORM)?.to_ptr() as cl_platform_id)
     }
 
+    #[cfg(feature = "CL_VERSION_1_2")]
```
This is a feature of an OpenCL device, not a CL_VERSION_1_2 OpenCL ICD feature.

kenba commented on src/device.rs:
```diff
@@ -400,54 +411,66 @@ impl Device {
         Ok(get_device_info(self.id(), DeviceInfo::CL_DEVICE_OPENCL_C_VERSION)?.to_string())
     }
 
+    #[cfg(feature = "CL_VERSION_1_2")]
```
As above, and sooo many below!

kenba commented on src/kernel.rs:
```diff
@@ -21,10 +21,9 @@ use super::Result;
 
 #[allow(unused_imports)]
 use cl3::ext;
-use cl3::types::{
```
Please use `#[allow(unused_imports)]` instead

kenba commented on src/kernel.rs:
```diff
@@ -168,6 +169,7 @@ impl Kernel {
         Ok(get_kernel_info(self.kernel, KernelInfo::CL_KERNEL_ATTRIBUTES)?.to_string())
     }
 
+    #[cfg(feature = "CL_VERSION_1_2")]
```
Not an OpenCL ICD feature as mentioned in other files.

kenba commented on src/kernel.rs:
```diff
@@ -168,6 +169,7 @@ impl Kernel {
     }
 
+    #[cfg(feature = "CL_VERSION_1_2")]
     pub fn get_arg_address_qualifier(&self, arg_indx: cl_uint) -> Result<cl_uint> {
```
As above and many below.

kenba commented on src/svm.rs:
```diff
@@ -18,7 +18,9 @@ use cl3::device::{
     CL_DEVICE_SVM_COARSE_GRAIN_BUFFER, CL_DEVICE_SVM_FINE_GRAIN_BUFFER,
     CL_DEVICE_SVM_FINE_GRAIN_SYSTEM,
 };
-use cl3::memory::{svm_alloc, svm_free, CL_MEM_READ_WRITE, CL_MEM_SVM_FINE_GRAIN_BUFFER};
```
Unnecessary change, disabled in `lib.rs`.

kenba commented on src/program.rs:
```diff
@@ -615,7 +620,7 @@ mod tests {
         assert_eq!(CL_PROGRAM_BINARY_TYPE_EXECUTABLE as u32, value);
 
-        // CL_VERSION_2_0 value
+        #[cfg(feature = "CL_VERSION_2_0")]
```
Unnecessary change, just getting info.

kenba reviewed (changes requested):
Sorry Volker, but I think that you have made a fundamental mistake about what is an OpenCL ICD function call and what information is available from different versions of OpenCL devices. This is particularly apparent in your changes to `device.rs`.

Please can you just make the minimum changes necessary to get the library working with your Mac ICD and *not* labelling functions (and tests) that are just calls to `cl*Info` functions with parameters for later versions of OpenCL?

vmx commented:
@kenba I understand what you mean. Though my idea was to follow what I find nice about Rust, which is that it let's you know about problems at compile time. Therefore I though it's nice that if you code for OpenCL 1.2 only and accidentally try to access OpenCL 2.0 properties, you will get a compile-time error. Why would you want to access OpenCL 2.0 features?

Though of course, if you prefer not having compile-time errors, that's fine with me.

vmx commented:
@kenba would it be OK if I add a document comment like `/// CL_VERSION_2_0` to public functions like `CommandQueue::size()` (instead of the compile time flag). Then it isn't enforced at compile-time, but people reading the docs would get that information without looking into the OpenCL spec.

kenba commented:
Yes Volker, that would be good.

vmx commented:
I think I've addressed all comments, I've a few notes:

 - I kept the tests that use SVM guarded by the `CL_VERSION_2_0` feature, I can change that if requested. 
 - I kept some methods in `Kernel`, e.g. `get_arg_address_qualifier()` behind the feature flag, as they use [`clGetKernelArgInfo`](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clGetKernelArgInfo), which is an OpenCL 1.2 feature.
 - I've made the feature flags consistent. I've removed some features in `Program`, e.g. from [`get_program_il`](https://github.com/kenba/opencl3/pull/31/files#diff-242e2cb3969431eb035a06581e743a67df53d7a1f2f1402ee8d88cd32f694725L456) as it only calls `get_program_info()`. It was interesting to see that there were cases that guarded the functions the way I did.
 - I've also made one unrelated change, in the current version it was not possible to compile the library with the `CL_VERSION_3_0` feature enabled. As I was testing it, I though I fix it right away.
 - Thanks for taking the time do do a thorough review. I hope that eventually my PRs will result in less work for you and not more. If there are any other changes needed, let me know or also feel free to fix them yourself.

vmx commented:
And yet again the confirmation that the PR still works on MacOS. I also have to admit that not having all "info" calls behind a feature flags makes my code simpler (though I still prefer explicit over implicit :)

kenba reviewed (approved):
Thanks Volker, it looks good now.
